Use std::span more in Modules/compression

cdumez opened the pull request:
#### 30df9b6d33e3c87ae61a76d71cd5709be08ee4f1
<pre>
Use std::span more in Modules/compression
<a href="https://bugs.webkit.org/show_bug.cgi?id=271467">https://bugs.webkit.org/show_bug.cgi?id=271467</a>

Reviewed by Darin Adler.

* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::decompress):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:

Canonical link: <a href="https://commits.webkit.org/276595@main">https://commits.webkit.org/276595@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18c95fe55e18f0833991461c6dac4d8aac5d5946

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47534 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47689 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47337 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21540 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36948 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45609 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21193 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18041 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18608 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39905 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3079 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41292 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40209 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49367 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16538 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43953 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21320 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42727 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21665 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6274 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20999 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
```diff
@@ -36,11 +36,10 @@ namespace WebCore {
-    auto* data = input.data();
-    if (!data)
+    if (!input.data())
         return Exception { ExceptionCode::TypeError, "No data provided"_s };
```
Is this really a reachable code path? Do we have test coverage? I’m kind of surprised that we have a “no data” case distinct from “data length zero”.

cdumez replied:
I don't think it is a distinct case? We could probably just check if the size is 0 here, I merely kept the existing code to be safe.

darinadler replied:
This won’t raise an exception for size 0 if somehow we make a `std::span` that has a non-null pointer for that case. I’d feel a lot more confident about this if we had test cases, and I didn’t check if we do. Not that refactoring to `std::span` affects this at all.

stwrt replied:
A BufferSource data() can return a nullptr, which is why I had added the check. An empty UInt8Array() being passed in is valid and has a size of 0, which is why we shouldn't add a check for length here.

There are no test cases that execute this code path from the ones I reviewed.

darinadler replied:
> A BufferSource data() can return a nullptr, which is why I had added the check.

The code below will do fine with a `nullptr` as long as the length is 0. It just won’t throw an exception. But why should it throw an exception? I don’t see how we can have an opinion about that unless we understand how it’s web-exposed.

> There are no test cases that execute this code path from the ones I reviewed.

We should construct the test case that will. Unless it’s impossible and this is really dead code, in which case we should simplify by removing the unneeded check.

stwrt replied:
I don't see a way to construct a test case that would trigger this exception.

I believe we can remove this check.

darinadler commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
```diff
@@ -34,11 +34,10 @@ namespace WebCore {
-    auto* data = input.data();
-    if (!data)
+    if (!input.data())
         return Exception { ExceptionCode::TypeError, "No data provided"_s };
```
Is this really a reachable code path? Do we have test coverage? I’m kind of surprised that we have a “no data” case distinct from “data length zero”.

darinadler commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -89,11 +89,11 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     bool m_usingAppleCompressionFramework { false };
 
-    inline ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompress(const uint8_t* input, const size_t inputLength);
+    inline ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompress(std::span<const uint8_t> input);
```
I don’t think we need the name `input` here.

darinadler commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -66,7 +66,7 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
     bool didInflateContainExtraBytes(int) const;
 
-    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressZlib(const uint8_t* input, const size_t inputLength);
+    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressZlib(std::span<const uint8_t> input);
```
I don’t think we need the name `input` here.

darinadler commented on Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:
```diff
@@ -89,11 +89,11 @@ class DecompressionStreamDecoder : public RefCounted<DecompressionStreamDecoder>
 #if PLATFORM(COCOA)
     compression_stream m_stream;
-    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressAppleCompressionFramework(const uint8_t* input, const size_t inputLength);
+    ExceptionOr<RefPtr<JSC::ArrayBuffer>> decompressAppleCompressionFramework(std::span<const uint8_t> input);
```
I don’t think we need the name `input` here.

cdumez committed:
Use std::span more in Modules/compression
https://bugs.webkit.org/show_bug.cgi?id=271467

Reviewed by Darin Adler.

* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::encode):
(WebCore::CompressionStreamEncoder::flush):
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/CompressionStreamEncoder.h:
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decode):
(WebCore::DecompressionStreamDecoder::flush):
(WebCore::DecompressionStreamDecoder::decompress):
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.h:

Canonical link: https://commits.webkit.org/276595@main
